File: compiler/src/parse/Names.swift
func parseTypeName(with parser: Parser) throws(ParseError) -> Substring {
    try parser.token(.capitalName, name: "a type name")
}

func parseConstructorName(with parser: Parser) throws(ParseError) -> Substring {
    try parser.token(.capitalName, name: "a constructor name")
}

func parseVariableName(with parser: Parser) throws(ParseError) -> Substring {
    try parser.token(.lowercaseName, name: "a variable name")
}

func parseTypeParameterName(with parser: Parser) throws(ParseError) -> Substring {
    try parser.token(.lowercaseName, name: "a type parameter name")
}

func parseAttributeName(with parser: Parser) throws(ParseError) -> Substring {
    if let name = try parser.parseOptional({ parser throws(ParseError) in
        try parser.token(.lowercaseName, name: "an attribute name")
    }) {
        return name
    }

    if let intrinsic = try parser.parseOptional({ parser throws(ParseError) in
        try parser.token(.intrinsicKeyword)
    }) {
        return intrinsic
    }

    throw parser.error("Expected an attribute name")
}

File: compiler/parse/src/names.rs
use crate::{
    lexer::TokenKind,
    parser::{ParseError, ParseToken, Parser},
};
use wipple_core::span::Str;

pub fn parse_type_name(parser: &mut Parser<'_>) -> Result<Str, ParseError> {
    parser.token(ParseToken::from(TokenKind::CapitalName).name("a type name"))
}

pub fn parse_constructor_name(parser: &mut Parser<'_>) -> Result<Str, ParseError> {
    parser.token(ParseToken::from(TokenKind::CapitalName).name("a constructor name"))
}

pub fn parse_variable_name(parser: &mut Parser<'_>) -> Result<Str, ParseError> {
    parser.token(ParseToken::from(TokenKind::LowercaseName).name("a variable name"))
}

pub fn parse_type_parameter_name(parser: &mut Parser<'_>) -> Result<Str, ParseError> {
    parser.token(ParseToken::from(TokenKind::LowercaseName).name("a type parameter name"))
}

pub fn parse_attribute_name(parser: &mut Parser<'_>) -> Result<Str, ParseError> {
    if let Some(name) = parser.parse_optional(|parser| {
        parser.token(ParseToken::from(TokenKind::LowercaseName).name("an attribute name"))
    })? {
        return Ok(name);
    }

    if let Some(intrinsic) =
        parser.parse_optional(|parser| parser.token(TokenKind::IntrinsicKeyword))?
    {
        return Ok(intrinsic);
    }

    Err(parser.error("Expected an attribute name", None))
}
